### DISCUSSION TITLE
Bash autocomplete?

### DISCUSSION THREAD

**Discussion by jason5122** [ImageMagick/ImageMagick#4608] [Help] [answered]
Hi, is there a way to get Bash autocomplete for commands? I've installed ImageMagick 7.1 via `brew` on macOS if that makes a difference. The usual way I see Homebrew programs do this is to put a script at `/usr/local/etc/bash_completion.d` (`git` does this for example), but I do not see one for ImageMagick.

**Comment by urban-warrior** [comment, marked as answer]
Create a Bash autocomplete configuration file for ImageMagick and post the link here.   We will modify the install script to add it to the `base_completion.d` folder.

**Reply by jason5122** [reply]
Okay, I'll look into it! I haven't done it before but hey, someone has to do it.

**Reply by jason5122** [reply]
So I did some nontrivial research into the subject and found a lot of useful guides/examples (I can post those here if anyone needs them). I think I can get it done, but I'm wondering if there is a good way to generate a list of command options and flags. I don't want to hardcode anything, like what is being done with the `convert` tool here: https://github.com/scop/bash-completion/blob/master/completions/convert.

Is this the right way to go about it? Should I scrape the `man` pages or the `--help` output?

**Comment by urban-warrior** [comment]
See https://imagemagick.org/script/command-line-options.php.  Does that help?  Otherwise `--help`.

**Comment by Jmuccigr** [comment]
FWIW my brew-installed versions autocompletes without any playing around.

**Reply by henrik-jensen** [reply]
Is that auto completion for ImageMagick or just for bash it self?
If the former I wonder where that elusive ImageMagick bash complete script comes from?
I would love to get my hands on that. What was your source for you brew installation?
Could I persuade you to check `$(brew --prefix)/etc/bash_completion.d/` (probably `/usr/local/etc/bash_completion.d/`. Confession: I know nothing about Home Brew or MacOs' implementation of the (bsd?) folder layout, but that's where the Internet hinted it might be) for an imagemagick related script, or maybe in your `~/.bashrc` or `~/.bash_profile` for an initialisation of such script? My guess is it would be called something like `im` or `imagemagick` perhaps with something like `-completion` appended.

**Comment by lukasz-r** [comment]
It would be really nice to have a Bash completion for the `magick` command. There is a completion for the deprecated `convert` command which works just fine, but the command is being dropped.

The completion script resides in
`/usr/share/bash-completion/completions/convert`
and is provided by the [Bash completion project](https://github.com/scop/bash-completion) which gathers all possible completion scripts.

Of course one could just copy the files, but the proper way to handle things is to write a good completion script adjusted to the new `magick` command (perhaps using the old completion script) and add it to the [Bash completion project](https://github.com/scop/bash-completion).

I've [submitted](https://github.com/ImageMagick/ImageMagick/issues/8692) it as a feature request.
